File: backend/FlashcardsApi/Services/UserService.cs
using FlashcardsApi.Models;
using System.Collections.Generic;
using System.Linq;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using System.IO;

namespace FlashcardsApi.Services
{
    public class UserService
    {
        private List<User> _users = new List<User>();
        private readonly string _filePath;

        public UserService()
        {
            _filePath = Path.Combine(AppContext.BaseDirectory, "Data", "users.json");

            // Load existing users if file exists
            if (File.Exists(_filePath))
            {
                try
                {
                    var json = File.ReadAllText(_filePath);
                    var loaded = JsonSerializer.Deserialize<List<User>>(json);
                    if (loaded != null)
                    {
                        _users.Clear();
                        _users.AddRange(loaded);
                    }
                }
                catch
                {
                    // ignore and start with empty list
                }
            }

            // Seed admin user if none exists
            if (!_users.Any())
            {
                var admin = new User
                {
                    Id = Guid.NewGuid().ToString(),
                    Username = "admin",
                    PasswordHash = HashPassword("admin123"),
                    Roles = new List<string> { UserRoles.Admin }
                };
                _users.Add(admin);
                Save();
            }
        }

        public IEnumerable<User> GetAll() => _users;
        public User GetById(string id) => _users.FirstOrDefault(u => u.Id == id);
        public User GetByUsername(string username) => _users.FirstOrDefault(u => u.Username == username);
        public void Add(User user)
        {
            _users.Add(user);
            Save();
        }
        public void Update(User user)
        {
            var idx = _users.FindIndex(u => u.Id == user.Id);
            if (idx >= 0) _users[idx] = user;
            Save();
        }
        public void Delete(string id)
        {
            _users.RemoveAll(u => u.Id == id);
            Save();
        }
        public bool ValidateCredentials(string username, string password)
        {
            var user = GetByUsername(username);
            return user != null && user.PasswordHash == HashPassword(password);
        }

        private void Save()
        {
            var dir = Path.GetDirectoryName(_filePath);
            if (!Directory.Exists(dir))
                Directory.CreateDirectory(dir!);

            var json = JsonSerializer.Serialize(_users, new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(_filePath, json);
        }
        public static string HashPassword(string password)
        {
            using var sha = SHA256.Create();
            var bytes = sha.ComputeHash(Encoding.UTF8.GetBytes(password));
            return Convert.ToBase64String(bytes);
        }
    }
}

File: backend/app/services/user_service.py
from __future__ import annotations
from typing import List, Optional
from ..models import User, UserSettings
import hashlib
import json
import uuid
import os

class UserService:
    """Service for managing users and their settings."""
    
    def __init__(self, path: str = "data/users.json"):
        self.path = path
        self._users: List[User] = []
        self._load()

    def _load(self):
        if os.path.exists(self.path):
            try:
                with open(self.path, "r", encoding="utf-8") as f:
                    data = json.load(f)
                    self._users = [User(**u) for u in data]
            except Exception:
                self._users = []
        if not self._users:
            admin = User(id=str(uuid.uuid4()), username="admin", password_hash=self.hash_password("admin123"), roles=["admin"])
            self._users.append(admin)
            self._save()

    def _save(self):
        os.makedirs(os.path.dirname(self.path), exist_ok=True)
        with open(self.path, "w", encoding="utf-8") as f:
            json.dump([u.dict() for u in self._users], f, indent=2)

    @staticmethod
    def hash_password(password: str) -> str:
        return hashlib.sha256(password.encode("utf-8")).hexdigest()

    def get_all(self) -> List[User]:
        return self._users

    def get_by_id(self, uid: str) -> Optional[User]:
        return next((u for u in self._users if u.id == uid), None)

    def get_by_username(self, username: str) -> Optional[User]:
        return next((u for u in self._users if u.username == username), None)

    def add(self, user: User):
        self._users.append(user)
        self._save()

    def update(self, user: User):
        for i, u in enumerate(self._users):
            if u.id == user.id:
                self._users[i] = user
                self._save()
                break

    def delete(self, uid: str):
        self._users = [u for u in self._users if u.id != uid]
        self._save()

    def validate_credentials(self, username: str, password: str) -> bool:
        user = self.get_by_username(username)
        if not user:
            return False
        return user.password_hash == self.hash_password(password)
